### PR TITLE
[orchagent]: Add MACsec Orchagent

### PR DESCRIPTION
**Pull request by Pterosaur**
**What I did**
Add MACsec orchagent for MACsec feature.
The MACsecOrch is introduced in the Orchagent to handle configuration requests. It monitors MACsec related tables in APP DB and convert those messages to SAI commands to manage the MACsec object. 
The main functions are defined in class MACsecOrch as follow
```
    task_process_status taskUpdateMACsecPort(const std::string & port_name, const TaskArgs & port_attr);
    task_process_status taskDisableMACsecPort(const std::string & port_name, const TaskArgs & port_attr);
    task_process_status taskUpdateEgressSC(const std::string & port_sci, const TaskArgs & sc_attr);
    task_process_status taskDeleteEgressSC(const std::string & port_sci, const TaskArgs & sc_attr);
    task_process_status taskUpdateIngressSC(const std::string & port_sci, const TaskArgs & sc_attr);
    task_process_status taskDeleteIngressSC(const std::string & port_sci, const TaskArgs & sc_attr);
    task_process_status taskUpdateEgressSA(const std::string & port_sci_an, const TaskArgs & sa_attr);
    task_process_status taskDeleteEgressSA(const std::string & port_sci_an, const TaskArgs & sa_attr);
    task_process_status taskUpdateIngressSA(const std::string & port_sci_an, const TaskArgs & sa_attr);
    task_process_status taskDeleteIngressSA(const std::string & port_sci_an, const TaskArgs & sa_attr);
```
 The HLD of MACsec orchagent is at [MACsec HLD](https://github.com/Azure/SONiC/blob/master/doc/macsec/MACsec_hld.md#344-macsec-orch)

**Why I did it**
MACsec orchagent is needed to transfer MACsec management message from wpa_supplicant to SAI.

**How I verified it**
Following the verification at https://github.com/Azure/sonic-swss/pull/1475 . The macsec device should be created and you can obverse whether the MACsec configuration is correct by command `ip macsec show`

**Details if related**
This PR depends on :
Azure/sonic-wpa-supplicant#16 
Azure/sonic-buildimage#5700
https://github.com/Azure/sonic-swss/pull/1475
https://github.com/Azure/sonic-sairedis/pull/691
https://github.com/Azure/sonic-swss-common/pull/434
https://github.com/Azure/sonic-sairedis/pull/770

### CODE REVIEW

**Comment by Pterosaur** [comment]
retest vs please

File: orchagent/port.h [hunk 1]
```diff
@@ -114,6 +114,8 @@ class Port
 
     std::unordered_set<sai_object_id_t> m_ingress_acl_tables_uset;
     std::unordered_set<sai_object_id_t> m_egress_acl_tables_uset;
+
+    sai_object_id_t     m_line_port_id = 0;
```
Comment (lguohan) [inline orchagent/port.h:118]: remove this one since this pr does not support gearbox

File: orchagent/macsecorch.h [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    /* Counter */
+    void installCounter(
+        CounterType counter_type,
+        const std::string &obj_name,
+        sai_object_id_t obj_id,
+        const std::vector<std::string> &stats);
+    void uninstallCounter(const std::string &obj_name, sai_object_id_t obj_id);
+
+    /* ACL */
+    bool initACLTable(
```
Comment (lguohan) [inline orchagent/macsecorch.h:203]: -> initMacsecAclTable

**Comment by Pterosaur** [comment]
Add test

**Comment by Pterosaur** [comment]
retest vs please

File: orchagent/macsecorch.cpp [hunk 3]
```diff
@@ -0,0 +1,2283 @@
...
+    {
+        "SAI_MACSEC_SA_ATTR_XPN",
+};
+
+static const std::vector<std::string> macsec_ingress_sa_stats =
+    {
+        "SAI_MACSEC_SA_ATTR_MINIMUM_XPN",
+};
+
+/* Helpers */
```
Comment (caizhenghui-juniper) [inline orchagent/macsecorch.cpp:38]: can helper functions be in swss-common?

File: orchagent/macsecorch.cpp [hunk 4]
```diff
@@ -0,0 +1,2283 @@
...
+                return nullptr;
+            }
+            auto &scs =
+                (m_direction == SAI_MACSEC_DIRECTION_EGRESS)
+                    ? port->m_egress_scs
+                    : port->m_ingress_scs;
+            auto sc = scs.find(*m_sci);
+            if (sc == scs.end())
+            {
+                SWSS_LOG_INFO("Cannot find the MACsec SC %lu at the port %s.", *m_sci, m_port_name->c_str());
```
Comment (qbdwlr) [inline orchagent/macsecorch.cpp:453]: Use PRIx64 instead of lu for SCI for easier debugging.

File: orchagent/macsecorch.cpp [hunk 5]
```diff
@@ -0,0 +1,2283 @@
...
+bool MACsecOrch::initMACsecObject(sai_object_id_t switch_id)
+{
+    SWSS_LOG_ENTER();
+
+    RecoverStack recover;
+
+    auto macsec_obj = m_macsec_objs.emplace(switch_id, MACsecObject());
+    if (!macsec_obj.second)
+    {
+        SWSS_LOG_INFO("The MACsec has been initialized at the switch %lu", switch_id);
```
Comment (qbdwlr) [inline orchagent/macsecorch.cpp:896]: Use PRIx64 instead of lu for consistent OID display in syslog.

File: orchagent/saihelper.cpp [hunk 6]
```diff
@@ -197,6 +199,7 @@ void initSaiApi()
     sai_log_set(SAI_API_SAMPLEPACKET,           SAI_LOG_LEVEL_NOTICE);
     sai_log_set(SAI_API_DEBUG_COUNTER,          SAI_LOG_LEVEL_NOTICE);
     sai_log_set((sai_api_t)SAI_API_NAT,         SAI_LOG_LEVEL_NOTICE);
+    sai_log_set(SAI_API_NAT,                    SAI_LOG_LEVEL_NOTICE);
```
Comment (qbdwlr) [inline orchagent/saihelper.cpp:202]: Copy-n-paste error.

File: orchagent/macsecorch.cpp [hunk 7]
```diff
@@ -0,0 +1,2283 @@
...
+#define MACSEC_STAT_FLEX_COUNTER_POLLING_INTERVAL_MS    (1000)
+#define COUNTERS_MACSEC_ATTR_GROUP                      "COUNTERS_MACSEC_ATTR"
+
+extern sai_object_id_t   gSwitchId;
+extern sai_macsec_api_t *sai_macsec_api;
+extern sai_acl_api_t *sai_acl_api;
+extern sai_port_api_t *sai_port_api;
+extern sai_switch_api_t *sai_switch_api;
+
+static const std::vector<std::string> macsec_egress_sa_stats =
```
Comment (qbdwlr) [inline orchagent/macsecorch.cpp:28]: When you add support for MACSEC_SA_STATS, this will need to be renamed since it is not a vector of stats but attrs.

**Review by qbdwlr — COMMENTED** [review]
Is this going to be committed before latest saimacsec.h changes are integrated?

File: orchagent/macsecorch.cpp [hunk 8]
```diff
@@ -0,0 +1,2283 @@
...
+        SWSS_LOG_WARN("Cannot create MACsec ingress port at the port %s", port_name.c_str());
+        return false;
+    }
+    recover.add_action([this, &macsec_port]() {
+        this->deleteMACsecPort(macsec_port.m_ingress_port_id);
+        macsec_port.m_ingress_port_id = SAI_NULL_OBJECT_ID;
+    });
+
+    macsec_port.m_enable_encrypt = true;
+    macsec_port.m_sci_in_sectag = true;
```
Comment (Pterosaur) [inline orchagent/macsecorch.cpp:1024]: SAI it's false by default.

**Review by caizhenghui-juniper — APPROVED** [review]

**Comment by Pterosaur** [comment]
retest vs please

**Review by lguohan — APPROVED** [review]

**Pull request merged**